File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans  = []
        path = []
        def dfs(nums,n,used,numUsed):
            if numUsed == n:
                ans.append(path[:])
                return 
            for i,num in enumerate(nums):
                if used[i]:
                    continue
                path.append(num)
                used[i] = True
                dfs(nums,n,used,numUsed+1)
                path.remove(num)
                used[i] = False
        n    = len(nums)
        used = [False] * n
        dfs(nums,n,used,0)
        return ans

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> ans;
    vector<int> path;
    void dfs(const vector<int>& nums, vector<int>& used, int n, int remaining){
        if (remaining == 0){
            ans.push_back(path);
            return;
        }
        for (int i=0; i<n; i++){
            if (used[i])
                continue;
            used[i] = 1;
            path.push_back(nums[i]);
            dfs(nums,used,n,remaining-1);
            path.pop_back();
            used[i] = 0;
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        int n = nums.size();
        vector<int> used(n,0);
        dfs(nums,used,n,n);
        return ans;
    }
};